services/ui/src/components/Problems/index.js

```diff
@@ -1,229 +1,368 @@
 
-const Problems = ({ problems }) => {
-    const { sortedItems, requestSort, getClassNamesFor } = useSortableProblemsData(problems);
+const getOptionsFromProblems = (problems, key) => {
+  let uniqueOptions= problems && new Set(problems.map(p => {
+    return p[key];
+  }));
+  return [...uniqueOptions];
+};
+
+const Problems = ({problems}) => {
+  const { sortedItems, requestSort, getClassNamesFor } = useSortableProblemsData(problems);
+  const [severitySelected, setSeverity] = useState([]);
+  const [sourceSelected, setSource] = useState([]);
+  const [servicesSelected, setService] = useState([]);
+
+  const [problemTerm, setProblemTerm] = useState('');
+  const [hasFilter, setHasFilter] = useState(false);
+  const [problemStats, setProblemStats] = useState([]);
+
+  const severities = getOptionsFromProblems(problems, 'severity');
+  const sources = getOptionsFromProblems(problems, 'source');
+  const services = getOptionsFromProblems(problems, 'service');
+
+
+  // Handlers
+  const handleProblemFilterChange = (event) => {
+    setHasFilter(false);
+
+    if (event.target.value !== null || event.target.value !== '') {
+      setHasFilter(true);
+    }
+    setProblemTerm(event.target.value);
+  };
+
+  const handleSourceChange = (source) => {
+    let values = source && source.map(s => s.value) || [];
+    setSource(values);
+  };
+
+  const handleSeverityChange = (severity) => {
+    let values = severity && severity.map(s => s.value) || [];
+    setSeverity(values);
+  };
+
+  const handleServiceChange = (service) => {
+    let values = service && service.map(s => s.value) || [];
+    setService(values);
+  };
+
+  // Options
+  const severityOptions = (severity) => {
+    return severity && severity.map(s => ({ value: s, label: s}));
+  };
+
+  const sourceOptions = (sources) => {
+    return sources && sources.map(s => ({ value: s, label: s}));
+  };
+
+  const serviceOptions = (services) => {
+    return services && services.map(s => ({ value: s, label: s}));
+  };
+
+  // Filters
+  const severityFilter = (item) => {
+    return Object.keys(item).some(key => {
+        if (item[key] !== null) {
+          return severitySelected.indexOf(item['severity'].toString()) > -1;
+        };
+    });
+  }
+
+  const sourceFilter = (item) => {
+    return Object.keys(item).some(key => {
+        if (item[key] !== null) {
+          return sourceSelected.indexOf(item['source'].toString()) > -1;
+        };
+    });
+  }
 
-    const [problemTerm, setProblemTerm] = useState('');
-    const [hasFilter, setHasFilter] = React.useState(false);
+  const serviceFilter = (item) => {
+    return Object.keys(item).some(key => {
+        if (item[key] !== null) {
+          return servicesSelected.indexOf(item['service'].toString()) > -1;
+        };
+    });
+  }
 
-    const handleProblemFilterChange = (event) => {
-        setHasFilter(false);
+  // Filter by select fields and text input field.
+  const filterResults = (item) => {
```

Comment:
The "filtered" logic here is a little confusing.

On line 97 you assign "filtered" true - and then you _check_ filtered on 99, 103, and 107 - but then you _also_ set filtered there. AND used filtered as a return value.

I think that the confusion mostly comes from the naming. The name of this function is "filterResults" which initially suggested to me that this should take some results, and filter them.
But after reading through the code, I realized that this isn't actually a filter, but it's a predicate function that tells us whether or not an _item should or should not be filtered_. This slightly confusing naming is repeated in the "sourceFilter", "severityFilter" and "serviceFilter" names.

These should be changed to something more accurate, I think - something that suggests that it's going to return a boolean value. (an isX() or shouldX() function name might make sense).

Reply:
Tidied up the naming of these now